# Week 5 Homework

#### Freanu Peria


### For this homework we will be using the FHV 2019-10 data found here. FHV Data

In [1]:
# download data
!wget https://github.com/DataTalksClub/nyc-tlc-data/releases/download/fhv/fhv_tripdata_2019-10.csv.gz

--2024-03-04 20:25:56--  https://github.com/DataTalksClub/nyc-tlc-data/releases/download/fhv/fhv_tripdata_2019-10.csv.gz
Resolving github.com (github.com)... 20.205.243.166
Connecting to github.com (github.com)|20.205.243.166|:443... connected.
HTTP request sent, awaiting response... 302 Found
Location: https://objects.githubusercontent.com/github-production-release-asset-2e65be/513814948/efdfcf82-6d5c-44d1-a138-4e8ea3c3a3b6?X-Amz-Algorithm=AWS4-HMAC-SHA256&X-Amz-Credential=AKIAVCODYLSA53PQK4ZA%2F20240304%2Fus-east-1%2Fs3%2Faws4_request&X-Amz-Date=20240304T122558Z&X-Amz-Expires=300&X-Amz-Signature=21c6ed11a15739c8908a307d1c826b75544fee29584689b0e3c2f7f701d68fab&X-Amz-SignedHeaders=host&actor_id=0&key_id=0&repo_id=513814948&response-content-disposition=attachment%3B%20filename%3Dfhv_tripdata_2019-10.csv.gz&response-content-type=application%2Foctet-stream [following]
--2024-03-04 20:25:57--  https://objects.githubusercontent.com/github-production-release-asset-2e65be/513814948/efdfcf82-6

In [3]:
# copy the file inside data folder
!mv fhv_tripdata_2019-10.csv.gz data/fhv

## Question 1:
## Install Spark and PySpark



Install Spark
Run PySpark
Create a local spark session
Execute spark.version.

In [15]:
import pyspark
from pyspark.sql import SparkSession
from pyspark.sql import types
from pyspark.sql import functions as F

import pandas as pd
pd.DataFrame.iteritems = pd.DataFrame.items # compatibility issues

In [5]:
spark = SparkSession.builder \
    .master("local[*]") \
    .appName('test') \
    .getOrCreate()

24/03/04 20:28:13 WARN Utils: Your hostname, coy-pc resolves to a loopback address: 127.0.1.1; using 172.19.138.153 instead (on interface eth0)
24/03/04 20:28:13 WARN Utils: Set SPARK_LOCAL_IP if you need to bind to another address


Setting default log level to "WARN".
To adjust logging level use sc.setLogLevel(newLevel). For SparkR, use setLogLevel(newLevel).


24/03/04 20:28:15 WARN NativeCodeLoader: Unable to load native-hadoop library for your platform... using builtin-java classes where applicable


In [7]:
spark.version

'3.3.2'

### Question 1 Answer:

3.3.2

## Question 2:
## FHV October 2019

Read the October 2019 FHV into a Spark Dataframe with a schema as we did in the lessons

Repartition the Dataframe to 6 partitions and save it to parquet

### What is the average size of the Parquet (ending with .parquet extension) Files that were created (in MB)? Select the answer which most closely matches.

In [ ]:
# First, we need to get the schema of fhv dataset. we can get the schema from a pandas dataframe

In [9]:
pd_fhv_df = pd.read_csv('data/fhv/fhv_tripdata_2019-10.csv.gz')

In [11]:
pd_fhv_df.head(5)

dispatching_base_num      pickup_datetime     dropOff_datetime  \
0               B00009  2019-10-01 00:23:00  2019-10-01 00:35:00   
1               B00013  2019-10-01 00:11:29  2019-10-01 00:13:22   
2               B00014  2019-10-01 00:11:43  2019-10-01 00:37:20   
3               B00014  2019-10-01 00:56:29  2019-10-01 00:57:47   
4               B00014  2019-10-01 00:23:09  2019-10-01 00:28:27   

   PUlocationID  DOlocationID  SR_Flag Affiliated_base_number  
0         264.0         264.0      NaN                 B00009  
1         264.0         264.0      NaN                 B00013  
2         264.0         264.0      NaN                 B00014  
3         264.0         264.0      NaN                 B00014  
4         264.0         264.0      NaN                 B00014

In [18]:
# create spark df from pd_fhv_df and get schema
# spark.createDataFrame(pd_fhv_df).schema
# tunrs out converting the pandas df to spark df causes an error because of the mixed dtypes that pandas have accepted.

# let's just create the schema
schema = types.StructType([
    types.StructField('dispatching_base_num', types.StringType(), True), 
    types.StructField('pickup_datetime', types.TimestampType(), True), 
    types.StructField('dropOff_datetime', types.TimestampType(), True), 
    types.StructField('PUlocationID', types.IntegerType(), True), 
    types.StructField('DOlocationID', types.IntegerType(), True), 
    types.StructField('SR_Flag', types.StringType(), True),
    types.StructField('Affiliated_base_number', types.StringType(), True)
    ])



In [20]:
fhv_df = spark.read \
    .option('header', 'true') \
    .schema(schema) \
    .csv('data/fhv/fhv_tripdata_2019-10.csv.gz')

In [21]:
fhv_df.show(n=5)

+--------------------+-------------------+-------------------+------------+------------+-------+----------------------+
|dispatching_base_num|    pickup_datetime|   dropOff_datetime|PUlocationID|DOlocationID|SR_Flag|Affiliated_base_number|
+--------------------+-------------------+-------------------+------------+------------+-------+----------------------+
|              B00009|2019-10-01 00:23:00|2019-10-01 00:35:00|         264|         264|   null|                B00009|
|              B00013|2019-10-01 00:11:29|2019-10-01 00:13:22|         264|         264|   null|                B00013|
|              B00014|2019-10-01 00:11:43|2019-10-01 00:37:20|         264|         264|   null|                B00014|
|              B00014|2019-10-01 00:56:29|2019-10-01 00:57:47|         264|         264|   null|                B00014|
|              B00014|2019-10-01 00:23:09|2019-10-01 00:28:27|         264|         264|   null|                B00014|
+--------------------+------------------

In [22]:
# repartition
fhv_df = fhv_df.repartition(6)

In [23]:
fhv_df.write.parquet("data/pq/fhv/201910")

In [25]:
!ls -lh data/pq/fhv/201910

total 39M
-rw-r--r-- 1 frean frean    0 Mar  4 20:42 _SUCCESS
-rw-r--r-- 1 frean frean 6.4M Mar  4 20:42 part-00000-9e3158f7-5cb3-44e0-9c8a-148b790a492a-c000.snappy.parquet
-rw-r--r-- 1 frean frean 6.4M Mar  4 20:42 part-00001-9e3158f7-5cb3-44e0-9c8a-148b790a492a-c000.snappy.parquet
-rw-r--r-- 1 frean frean 6.4M Mar  4 20:42 part-00002-9e3158f7-5cb3-44e0-9c8a-148b790a492a-c000.snappy.parquet
-rw-r--r-- 1 frean frean 6.4M Mar  4 20:42 part-00003-9e3158f7-5cb3-44e0-9c8a-148b790a492a-c000.snappy.parquet
-rw-r--r-- 1 frean frean 6.4M Mar  4 20:42 part-00004-9e3158f7-5cb3-44e0-9c8a-148b790a492a-c000.snappy.parquet
-rw-r--r-- 1 frean frean 6.4M Mar  4 20:42 part-00005-9e3158f7-5cb3-44e0-9c8a-148b790a492a-c000.snappy.parquet


### Question 2 Answer:

6.4Mb

## Question 3
## Count records

##### How many taxi trips were there on the 15th of October?

##### Consider only trips that started on the 15th of October.

In [32]:
fhv_df.show(n=5)

+--------------------+-------------------+-------------------+------------+------------+-------+----------------------+
|dispatching_base_num|    pickup_datetime|   dropOff_datetime|PUlocationID|DOlocationID|SR_Flag|Affiliated_base_number|
+--------------------+-------------------+-------------------+------------+------------+-------+----------------------+
|              B02111|2019-10-27 05:35:21|2019-10-27 05:38:04|          92|          80|   null|                B02111|
|              B01437|2019-10-08 09:41:27|2019-10-08 09:46:52|         264|         197|   null|                B01437|
|              B02107|2019-10-09 16:53:55|2019-10-09 17:06:05|         264|         167|   null|                B02107|
|              B01653|2019-10-15 06:47:13|2019-10-15 06:59:36|         264|         264|   null|                B01653|
|              B00850|2019-10-13 03:57:31|2019-10-13 04:12:36|         264|         177|   null|                B00850|
+--------------------+------------------

In [28]:
# I would like to use SQL on this.
# so I will be making a view on our spark session

fhv_df.createOrReplaceTempView('fhv_data')

In [51]:
oct_trips_res = spark.sql("""
    SELECT COUNT(1)
    FROM fhv_data
    WHERE pickup_datetime >= '2019-10-15 00:00:00'
    and dropoff_datetime <= '2019-10-15 23:59:59'
    --and dropoff_datetime <= '2019-10-16'
""")

In [52]:
oct_trips_res.show()

+--------+
|count(1)|
+--------+
|   61851|
+--------+



### Question 3 Answer:

62, 610 (closest answer)

## Question 4
## Longest trip for each day

What is the length of the longest trip in the dataset in hours?

In [84]:
# let's try to add a new column which calculates difference of pickup from dropoff
# good thing i found a solution on how to subtract timestamps
# from https://sparkbyexamples.com/pyspark/pyspark-timestamp-difference-seconds-minutes-hours/

fhv_df.withColumn(
    'trip_duration', 
    (F.col('pickup_datetime').cast('long') - F.col('dropoff_datetime').cast('long')) / 3600
    ) \
    .orderBy('trip_duration') \
    .show(n=10)

+--------------------+-------------------+-------------------+------------+------------+-------+----------------------+-------------------+
|dispatching_base_num|    pickup_datetime|   dropOff_datetime|PUlocationID|DOlocationID|SR_Flag|Affiliated_base_number|      trip_duration|
+--------------------+-------------------+-------------------+------------+------------+-------+----------------------+-------------------+
|              B02832|2019-10-11 18:00:00|2091-10-11 18:30:00|         264|         264|   null|                B02832|          -631152.5|
|              B02832|2019-10-28 09:00:00|2091-10-28 09:30:00|         264|         264|   null|                B02832|          -631152.5|
|              B02416|2019-10-31 23:46:33|2029-11-01 00:13:00|        null|        null|   null|                B02416| -87672.44083333333|
|     B00746         |2019-10-01 21:43:42|2027-10-01 21:45:23|         159|         264|   null|       B00746         | -70128.02805555555|
|              B0292

### Question 4 Answer

631,152.5 Hours (^First row in above, with dispatching_base_num=B02832)

## Question 5
## User Interface

Spark’s User Interface which shows the application's dashboard runs on which local port?
Answer: 4040


##  Question 6:
## Least frequent pickup location zone

Using the zone lookup data and the FHV October 2019 data, what is the name of the LEAST frequent pickup location Zone?

In [86]:
# this is already taxi_zone_lookup. we can read it in our data.

zones_df = spark.read \
    .option("header", "true") \
    .csv('taxi+_zone_lookup.csv')

In [87]:
zones_df.show(n=5)

+----------+-------------+--------------------+------------+
|LocationID|      Borough|                Zone|service_zone|
+----------+-------------+--------------------+------------+
|         1|          EWR|      Newark Airport|         EWR|
|         2|       Queens|         Jamaica Bay|   Boro Zone|
|         3|        Bronx|Allerton/Pelham G...|   Boro Zone|
|         4|    Manhattan|       Alphabet City| Yellow Zone|
|         5|Staten Island|       Arden Heights|   Boro Zone|
+----------+-------------+--------------------+------------+
only showing top 5 rows



In [88]:
# I would like to use SQL again :D
# lets create another view
zones_df.createOrReplaceTempView('zones')

In [94]:
least_frequent_zones_df = spark.sql("""

    SELECT COUNT(1), b.zone
    FROM fhv_data a
        JOIN zones b
        ON a.PUlocationID = b.locationID
    GROUP BY b.zone
    ORDER BY COUNT(1)
""")

In [96]:
least_frequent_zones_df.show(n=5)

+--------+--------------------+
|count(1)|                zone|
+--------+--------------------+
|       1|         Jamaica Bay|
|       2|Governor's Island...|
|       5| Green-Wood Cemetery|
|       8|       Broad Channel|
|      14|     Highbridge Park|
+--------+--------------------+
only showing top 5 rows



### Question 6 Answer

Jamaica Bay
